.travis.yml

```diff
@@ -2,22 +2,76 @@
 language: c++
 sudo: required
 
+services:
+  - docker
+
 os:
   - linux
   - osx
 
-dist: trusty # Ubuntu 14.04.
+env:
+  global:
+    - BUILD_TYPE=Debug
+    - TECA_DIR=/travis_teca_dir
+    - TECA_DATA_REVISION=39
+  matrix:
+    - DOCKER_IMAGE=ubuntu IMAGE_VERSION=18.04 SCRIPT_DIR=ubuntu_18_04
+    - DOCKER_IMAGE=fedora IMAGE_VERSION=28 SCRIPT_DIR=fedora_28
+    - NO_DOCKER=TRUE
 
-compiler:
-  - gcc
+matrix:
+  exclude:
+    - os: osx
+      env: DOCKER_IMAGE=ubuntu IMAGE_VERSION=18.04 SCRIPT_DIR=ubuntu_18_04
+    - os: osx
+      env: DOCKER_IMAGE=fedora IMAGE_VERSION=28 SCRIPT_DIR=fedora_28
+    - os: linux
+      env: NO_DOCKER=TRUE
 
-install: ./test/travis_ci/install_$TRAVIS_OS_NAME.sh
+before_install:
+  - >
+    if [[ "$TRAVIS_OS_NAME" != "osx" ]];
+    then
+    docker pull $DOCKER_IMAGE:$IMAGE_VERSION &&
+    docker run -t -v $TRAVIS_BUILD_DIR:$TECA_DIR -w $TECA_DIR
+    --name teca_${DOCKER_IMAGE}_${IMAGE_VERSION} -d $DOCKER_IMAGE:$IMAGE_VERSION;
+    fi
 
-env:
-  - BUILD_TYPE=Debug
-#  - BUILD_TYPE=Release
+install:
+  - >
+    if [[ "$TRAVIS_OS_NAME" != "linux" ]];
+    then
+    ./test/travis_ci/install_$TRAVIS_OS_NAME.sh;
+    elif [[ "$TRAVIS_OS_NAME" != "osx" ]];
+    then
+    docker exec teca_${DOCKER_IMAGE}_${IMAGE_VERSION} /bin/bash -c
+    "$TECA_DIR/test/travis_ci/docker/$SCRIPT_DIR/install_$DOCKER_IMAGE.sh";
+    fi
 
-script: ./test/travis_ci/ctest_$TRAVIS_OS_NAME.sh
+script:
+  - >
+    if [[ "$TRAVIS_OS_NAME" != "linux" ]];
+    then
+    ./test/travis_ci/ctest_$TRAVIS_OS_NAME.sh;
+    elif [[ "$TRAVIS_OS_NAME" != "osx" ]];
+    then
+    docker exec teca_${DOCKER_IMAGE}_${IMAGE_VERSION} /bin/bash -c
+    "export TRAVIS_BRANCH=$TRAVIS_BRANCH &&
+    export BUILD_TYPE=$BUILD_TYPE &&
+    export DOCKER_IMAGE=$DOCKER_IMAGE &&
+    export IMAGE_VERSION=$IMAGE_VERSION &&
+    $TECA_DIR/test/travis_ci/docker/ctest_linux.sh";
+    fi
 
 after_failure:
-  - cat build/Testing/Temporary/LastTest.log
+  - >
+    if [[ "$TRAVIS_OS_NAME" != "linux" ]];
+    then
+    cat build/Testing/Temporary/LastTest.log;
```

Comment:
fix indentation, here and elsewhere shell script should be indented with 4 spaces

Reply:
Yeah I tried this but that format doesn't work with yaml.  I tried even their [indentation feature](https://yaml.org/spec/1.2/spec.html#id2793979) and it didn't work because it requires all the lines to be equally indented.

Reply:
OK

Reply:
OK. I guess we have to live with it, or move shell code into a script. If you want to go for it! If not, this is fine.